README.md

```diff
@@ -11,4 +11,29 @@
 ## Homework 02
  - Write an echo module, which recieves :ping atom as a message and returns {:pong, node()} tuple.
  - Write unit tests.
- 
\ No newline at end of file
+ 
+## Homework 04
+ The project is the implementation of tram [finite state machine](https://en.wikipedia.org/wiki/Finite-state_machine) on pure Elixir language according the following diagram:
+
+```renderAs=mermaid
+flowchart TD
+  Depot ---> |start the route| Move_to_station
+  Move_to_station ---> |finish the route| Depot
+  Move_to_station ---> |arrive to station| Stop_on_station
+  Stop_on_station ---> |moving to next station| Move_to_station
+  Stop_on_station ---> |open the doors| Doors_opened
+  Doors_opened ---> |close the doors|Stop_on_station
+  Doors_opened ---> |start onboarding passengers| Onboarding_passengers
+  Onboarding_passengers ---> |stop onboarding passengers| Doors_opened
+  Move_to_station ---> |railway_blocked| Stop_by_railway_block
+  Stop_by_railway_block ---> |continue the route| Move_to_station
+  Move_to_station ---> |traffic light red| Stop_by_traffic_light
+  Stop_by_traffic_light ---> |traffic light green| Move_to_station
+  Move_to_station ---> |unexpected accident happened| Stop_by_accident
+  Stop_by_accident ---> |open the doors| Emergency_doors_opening
+  Emergency_doors_opening --->|start rescue passengers| Rescue_passengers
+  Rescue_passengers ---> |stop rescue passengers| Emergency_doors_opening
+  Emergency_doors_opening ---> |close the doors| Stop_by_accident
+  Stop_by_accident ---> |move to service station| Service
+  Service ---> |back to depot| Depot
```

Comment:
OMG

You don’t like gradual approaches, do you? :) A bit of general advice, while I didn’t get further to read the implementation: start with implementing “main” states, which are 5 tops, then explore each of them. Einstein supposedly said (well, this is usually attributed to him,) that the normal human can operate 5 different entities at a time, and for those geniuses, this number is 7. Don’t prove the humanity is stupid by inventing the code requiring readers to manipulate 10+ :)

---

Also, I’d say you sometimes confuse _tram_ states and their internal attributes. What about keeping _whys_ in attributes and _whats_ in the state itself? Like (I am oversimplifying it, but still)
```elixir
in_depot → |move| moving
moving → |stop| on_stop # conditional
moving → |bang| on_accident
on_stop → |doors| on_stop # conditional
on_stop → |move| moving # this checks if doors are closes etc
on_stop → |move| to_depot # final state
on_accident → |doors| on_accident
on_accident → |repair| moving
```
Notice `move` is _the very same event_ accepted in two states (`in_depot` and `on_stop`) _and_ resulting in two states (`moving`, `to_depot`.)

Reply:
Yep, 5±2 numbers are also called "Miller's wallet". Valid point. I'll try to reduce reading complexity of my code in the next iteration. :)